### PR TITLE
Add proper surrogate key headers for API users#show, articles#feed, stories#index (for users), twitch#show

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
This adds edge caching to the `/api/users/:id` endpoint which will make it more scalable. But adding that means we need to keep track of its caching lifecycle effectively. So this also adds _proper_ surrogate control headers building on #4744.

### CODE REVIEW

File: app/labor/cache_buster.rb [hunk 1]
```diff
@@ -172,6 +172,11 @@ def self.bust_user(user)
       "/feed/#{username}"
     ]
     paths.each { |path| bust(path) }
+
+    # ^ TODO We can remove the above most or all of the manual paths purge
+    # once past relevant caches expire and get new surrogate keys (> 1 day after merge).
+
+    user.purge
```
Comment (mstruve) [inline app/labor/cache_buster.rb]: 🎉

File: app/controllers/stories_controller.rb [hunk 2]
```diff
@@ -176,7 +176,7 @@ def handle_user_index
     redirect_if_view_param
     return if performed?
 
-    set_surrogate_key_header "articles-user-#{@user.id}"
+    set_surrogate_key_header @user.record_key
```
Comment (rhymes) [inline app/controllers/stories_controller.rb]: What do you think about making this more flexible, we should at least add `@stories` record keys to this as this returns a collection of articles.

What about:

```ruby
set_surrogate_key_header @user.record_key, "articles", @stories.map(&:record_key)
```

?

The article purging code should get in the code if we merge https://github.com/thepracticaldev/dev.to/pull/5606

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -23,10 +23,8 @@ def feed
       render body: nil
       return
     end
-
-    set_surrogate_key_header "feed"
-    response.headers["Surrogate-Control"] = "max-age=600, stale-while-revalidate=30, stale-if-error=86400"
-
+    set_surrogate_keys_by_case
+    response.headers["Surrogate-Control"] = "max-age=186400, stale-while-revalidate=3600, stale-if-error=86400"
```
Comment (mstruve) [inline app/controllers/articles_controller.rb]: Today I learned what all these headers mean: https://www.fastly.com/blog/stale-while-revalidate-stale-if-error-available-today

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -244,6 +243,16 @@ def set_article
     @article = found_article || not_found
   end
 
+  def set_surrogate_keys_by_case
+    if @tag_model
+      set_surrogate_key_header @tag_model.record_key, Article.table_key, Tag.table_key, @articles.map(&:record_key)
+    elsif @user
+      set_surrogate_key_header @user.record_key, Article.table_key, User.table_key, @articles.map(&:record_key)
+    else
```
Comment (mstruve) [inline app/controllers/articles_controller.rb]: Do we need a case for organization here?

File: app/controllers/articles_controller.rb (hunk 3 above)
Comment (rhymes) [inline app/controllers/articles_controller.rb]: I haven't noticed so far but we're calling setting caching twice I think we should use declarative programming and keep using `:set_cache_control_headers` which does exactly that:

https://github.com/fastly/fastly-rails/blob/6ed13a694b8753d43515079ae667fe6f8c0fcd23/lib/fastly-rails/action_controller/cache_control_headers.rb#L10-L12

So, as the pattern around the code base is "set cache control headers in a before action and then set surrogate keys at the end of the controller", let's do it like this:

```ruby
before_action -> { set_cache_control_headers(186_400, stale_while_revalidate: 30, stale_if_error: 86_400) }, only: %i[feed]
```

or even better (personally I end up googling "186400 seconds in hours" everytime I see these numbers :D) we can do:

```ruby
  before_action only: %i[feed] do
    set_cache_control_headers(51.hours + 46.minutes + 40.seconds, stale_while_revalidate: 1.hour, stale_if_error: 24.hours)
  end
```


as we're already calling set cache control headers.

I think it's going to be clearer to the reader as how long we're caching our RSS feed for.

We might also want to add a test or not. 

BTW I've noticed, by testing this manually locally, that we're serving our feed with the service worker's shell. I've sent a PR to fix that: https://github.com/thepracticaldev/dev.to/pull/5734

File: app/controllers/articles_controller.rb (hunk 4 above)
Comment (rhymes) [reply app/controllers/articles_controller.rb]: Yep, I think so, orgs have their feed: https://dev.to/feed/devteam

File: app/controllers/articles_controller.rb [hunk 5]
```diff
@@ -244,6 +243,16 @@ def set_article
     @article = found_article || not_found
   end
 
+  def set_surrogate_keys_by_case
```
Comment (rhymes) [inline app/controllers/articles_controller.rb]: I don't think we should have this method at all.

We already have a `if` in the `feed` which either calls `handle_user_or_organization_feed` or `handle_tag_feed` or `@articles.where(featured: true).includes(:user)`, let's put the proper set_surrogate_key call at the end of those methods. This way we don't have to wonder where these instance variables come from. For example:

```ruby
  def handle_tag_feed
    @tag = Tag.aliased_name(params[:tag])
    return unless @tag

    @articles = @articles.cached_tagged_with(@tag)
    
    set_surrogate_key_header @tag_model.record_key, Article.table_key, Tag.table_key, @articles.map(&:record_key)
  end
```

**Comment by mstruve** [comment]
@rhymes can we close this? I think you fixed all these issues when you were overhauling our surrogate headers not long ago.

**Comment by rhymes** [comment]
@mstruve think so, also @atsmith813 is overhauling caching as well. If we need more we can add them case by case instead of a big PR.

**Pull request closed without merging**